src/osd/scrubber/osd_scrub_sched.h

```diff
@@ -4,550 +4,713 @@
+  int deep_randomize_pcnt{0};
 
-    /*
-     * as the atomic 'in_queues' appears in many log prints, accessing it for
-     * display-only should be made less expensive (on ARM. On x86 the _relaxed
-     * produces the same code as '_cs')
-     */
-    std::string_view registration_state() const
-    {
-      return in_queues.load(std::memory_order_relaxed) ? "in-queue"
-						       : "not-queued";
-    }
+  /// must we schedule a scrub with high urgency if we do not have a valid
+  /// last scrub stamp?
+  bool mandatory_on_invalid{true};
+};
 
-    /**
-     * a text description of the "scheduling intentions" of this PG:
-     * are we already scheduled for a scrub/deep scrub? when?
-     */
-    std::string scheduling_state(utime_t now_is, bool is_deep_expected) const;
+class ScrubJob;
 
-    friend std::ostream& operator<<(std::ostream& out, const ScrubJob& pg);
-  };
+/*
+ * There are two versions of the scheduling-target (the object detailing one
+ * of the two scrub types (deep or shallow) for a specific PG):
+ * 'SchedEntry' is maintained by the ScrubQueue, and holds just
+ *   the scheduling details.
+ * 'SchedTarget' is maintained by the PgScrubber, and
+ *   holds the same set of scheduling details plus additional information
+ *   about the scrub to be performed.
+ */
 
-  friend class TestOSDScrub;
-  friend class ScrubSchedTestWrapper; ///< unit-tests structure
+struct SchedEntry {
+  static inline constexpr const utime_t eternity{
+      time_t{std::numeric_limits<int32_t>::max()}, 0};
 
-  using ScrubJobRef = ceph::ref_t<ScrubJob>;
-  using ScrubQContainer = std::vector<ScrubJobRef>;
+  constexpr SchedEntry(spg_t pgid, scrub_level_t level)
+      : pgid{pgid}
+      , level{level}
+  {}
 
-  static std::string_view qu_state_text(qu_state_t st);
+  spg_t pgid;
+  scrub_level_t level;
 
-  /**
-   * called periodically by the OSD to select the first scrub-eligible PG
-   * and scrub it.
-   *
-   * Selection is affected by:
-   * - time of day: scheduled scrubbing might be configured to only happen
-   *   during certain hours;
-   * - same for days of the week, and for the system load;
-   *
-   * @param preconds: what types of scrub are allowed, given system status &
-   *                  config. Some of the preconditions are calculated here.
-   * @return Scrub::attempt_t::scrubbing if a scrub session was successfully
-   *         initiated. Otherwise - the failure cause.
-   *
-   * locking: locks jobs_lock
-   */
-  Scrub::schedule_result_t select_pg_and_scrub(Scrub::ScrubPreconds& preconds);
+  /// 'white-out' support: if false, the entry was logically removed from
+  /// the queue
+  bool is_valid{true};
+
+  urgency_t urgency{urgency_t::off};
 
   /**
-   * Translate attempt_ values into readable text
+   * the time at which we are allowed to start the scrub. Never
+   * decreasing after 'target' is set.
    */
-  static std::string_view attempt_res_text(Scrub::schedule_result_t v);
+  utime_t not_before{eternity};
```

Comment:
All the time members here default to "way in the future", so that they compare correctly. For the NB specifically - one is not expected to see an entry with 'target' or 'urgency' set while the NB was not, but I prefer the uniformity.